README.md
<p align=center>
  <a href="https://github.com/TimoKunze/StatusBarControl/releases/tag/1.5.4">
    <img alt="Release 1.5.4 Release" src="https://img.shields.io/badge/release-1.5.4-0688CB.svg">
  </a>
  <a href="https://github.com/TimoKunze/StatusBarControl/releases">
    <img alt="Download StatusBarControl" src="https://img.shields.io/badge/download-latest-0688CB.svg">
  </a>
  <a href="https://github.com/TimoKunze/StatusBarControl/blob/master/LICENSE">
    <img alt="License" src="https://img.shields.io/badge/license-MIT-0688CB.svg">
  </a>
  <a href="https://www.paypal.com/xclick/business=TKunze71216%40gmx.de&item_name=StatusBarControl&no_shipping=1&tax=0&currency_code=EUR">
    <img alt="Donate" src="https://img.shields.io/badge/%24-donate-E44E4A.svg">
  </a>
</p>

# StatusBarControl
An ActiveX control for Visual Basic 6 that can be used as a status bar.

I've developed this ActiveX control in 2005 and did update it on a regular basis until 2016. Currently I have little interest to maintain this project any longer, but I think the code might be of some use to others.

# Before you make changes
If you make changes to the code and deploy the binary, keep in mind that ActiveX controls are COM components and therefore should stay binary compatible as long as you don't change the COM object's, i.e. the ActiveX control's public class name and GUIDs. Otherwise people using these components are likely to end up in the famous COM hell.

# Dependencies
You will need the [Microsoft Windows 10 SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk), ATL, and [WTL 10](https://sourceforge.net/projects/wtl/).

## Required Change to ATL
Some version of ATL have a bug in ```AtlIPersistPropertyBag_Load``` which causes crashes. In the file atlcom.h search for ```AtlIPersistPropertyBag_Load```. Inside the implementation of this function search for ```HRESULT hr = pPropBag->Read(pMap[i].szDesc, &var, pErrorLog);```. Make sure that there is this code before this line: ```var.pdispVal = NULL;```.

CHANGELOG
Version 1.5.4 - 04/02/2018

- Fixed multiple memory leaks.
- Upgraded to WTL 10.
- Integrated the GitHub repository into the documentation and the About dialog.



Version 1.5.3 - 15/01/2018

- Project published on GitHub.



Version 1.5.2 - 31/07/2016

- Fixed some memory leaks that occurred when firing certain events.



Version 1.5.1 - 30/01/2016

- FIX: The control could crash during OLE drag'n'drop.
- FIX: The control could crash the other process during OLE drag'n'drop, especially when IOLEDataObject::SetDropDescription() had been called.



Version 1.5.0 - 23/01/2016

- Added the IStatusBar::FinishOLEDragDrop method. It should not be used if not really needed. It can be used to hide the drag image if execution of the OLEDragDrop event takes a long time or if a message is displayed from within this event.
- Initial support for Windows 10.



Version 1.4.0 - 17/07/2014

- Added the XClick, and XDblClick events for support of the extended mouse buttons.
- Extended the Mouse* events to support the extended mouse buttons.
- FIX: The Font and MouseIcon properties could leak GDI resources.
- Initial support for Windows 8.1.



Version 1.3.3 - 27/03/2013

- FIX: The ResizedControlWindow event was not fired when it should have been fired.
- Initial support for Windows 8.



Version 1.3.2 - 03/04/2012

- FIX: Clicking the control no longer raises the Validate event of the currently focused control.



Version 1.3.1 - 21/05/2011

- Fixed a bug that could make other, invisible windows visible again under rather complex circumstances. I'm not sure this could happen practically, but in theory it could happen and practically it did happen for the Frame control which had the same bug.
- FIX: The ResizedControlWindow event was fired when actually nothing had changed.



Version 1.3.0 - 08/01/2011

- FIX: The control could crash when it was destroyed.
- Switched to Visual C++ 2010 and dropped support for Windows 2000.



Version 1.2.3 - 18/10/2010

- FIX: The control could crash with an access violation while trying to display the tool tip text. This was caused by a bug in comctl32.dll that now is being worked around.



Version 1.2.2 - 05/07/2010

- Added a new property page that can be used to enter Unicode texts at design time.
- FIX: IOLEDataObject::SetDropDescription could cause crashes in other programs.



Version 1.2.1 - 02/01/2010

- Improved support for Aero drag images.
- Improved performance.
- FIX: The 'shift' parameter that many events have, often missed the ALT key.
- Deploy (stripped) symbols with the binaries to make debugging easier. This are the pdb files in case you wonder.



Version 1.2.0 - 25/07/2009

- Initial support for Windows 7.
- Extended the DropDescriptionIconConstants enumeration by 'ddiUseDefault'.
- FIX: Drop descriptions didn't work as expected if the drag source was a 64 bit app.
- Massively reduced the usage of GDI handles.



Version 1.1.1 - 30/10/2008

- FIX: Displayed key states could be wrong.
- FIX: The last panel could be drawn wrong if its content was not left-aligned.
- FIX: Ignore calls to IOLEDataObject::SetDropDescription() if the drag image is in classic style.
- FIX: Changing the IStatusBar::BackColor property could fail if its current value was -1.
- Fixed some GDI handle leaks.
- Fixed some drawing glitches.
- Slightly changed handling of the situation where the user presses a mouse button, drags the mouse out of the control's client area and releases the button.
- Removed the data-binding stuff (it was probably broken anyway).
- Switched to Visual C++ 2008 and dropped support for Windows 9x, Me and NT 4.0.
- Improved the documentation.



Version 1.1.0 - 04/12/2007

- Added the IOLEDataObject::GetDropDescription() method.
- Added the IOLEDataObject::SetDropDescription() method.
- Removed the limitation to 127 characters for panel texts for comctl32.dll version 6.10 or higher.
- FIX: Changes were not applied implicitly when jumping between multiple instances of the control with the property pages opened.
- Improved the documentation.



Version 1.0.6 - 28/02/2007

- Decreased the binaries' size.
- Improved the documentation.
- Ported the WTL samples to WTL 8.0.
- Made the samples DPI aware.
- Linked the new bug tracking system.



Version 1.0.5 - 23/09/2006

- FIX: The "Font" and "MouseIcon" properties didn't work as expected.
- Improved Vista compatibility.
- Improved the property pages.
- Improved setup.



Version 1.0.4 - 11/06/2006

- FIX: Changing the font could crash the control.
- FIX: Moving the control without resizing it could lead to a control width or height of 0.
- Fixed some repaint problems.
- Improved IStatusBar::Font and IStatusBar::MouseIcon.
- Improved IStatusBar::LoadSettingsFromFile().
- Changed the control's default size to 100x23.
- Minor bugfixes.



Version 1.0.3 - 18/12/2005

- Switched to Visual C++ 2005; fixed compiler errors and warnings.
- FIX: The control crashed if it was destroyed from within an event handler.
- FIX: The IStatusBar::DontRedraw property interfered with the control's Visible property.
- Ported the WTL7 samples to Visual C++ 2005.



Version 1.0.2 - 14/11/2005

- Added some C++/WTL samples.
- FIX: The behavior of the IStatusBar::Enabled property was different from other controls.
- Improved Drag'n'Drop support.
- Improved documentation.



Version 1.0.1 - 28/09/2005

- Renamed the _IStatusBarEvents::DestroyedControl event to _IStatusBarEvents::DestroyedControlWindow.
- Renamed the _IStatusBarEvents::RecreatedControl event to _IStatusBarEvents::RecreatedControlWindow.
- Renamed the _IStatusBarEvents::ResizedControl event to _IStatusBarEvents::ResizedControlWindow.
- Improved control resizing performance.
- Revised the Font property.
- Revised documentation.
- FIX: The Font property leaked GDI objects.
- FIX: The sample projects crashed if you clicked the "Show property pages..." button and selected the Picture property page.

src/DispIDs.h
// DispIDs.h: Defines a DISPID for each COM property and method we're providing.

// IStatusBar
// properties
#define DISPID_STATBAR_APPEARANCE													1
#define DISPID_STATBAR_BACKCOLOR													2
#define DISPID_STATBAR_BORDERSTYLE												3
#define DISPID_STATBAR_CUSTOMCAPSLOCKTEXT									4
#define DISPID_STATBAR_CUSTOMINSERTKEYTEXT								5
#define DISPID_STATBAR_CUSTOMKANALOCKTEXT									6
#define DISPID_STATBAR_CUSTOMNUMLOCKTEXT									7
#define DISPID_STATBAR_CUSTOMSCROLLLOCKTEXT								8
#define DISPID_STATBAR_DISABLEDEVENTS											9
#define DISPID_STATBAR_DONTREDRAW													10
#define DISPID_STATBAR_ENABLED														DISPID_ENABLED
#define DISPID_STATBAR_FONT																11
#define DISPID_STATBAR_FORCESIZEGRIPPERDISPLAY						12
#define DISPID_STATBAR_HOVERTIME													13
#define DISPID_STATBAR_HWND																14
#define DISPID_STATBAR_MINIMUMHEIGHT											15
#define DISPID_STATBAR_MOUSEICON													16
#define DISPID_STATBAR_MOUSEPOINTER												17
#define DISPID_STATBAR_PANELS															18
#define DISPID_STATBAR_PANELTOAUTOSIZE										19
#define DISPID_STATBAR_REGISTERFOROLEDRAGDROP							20
#define DISPID_STATBAR_RIGHTTOLEFTLAYOUT									21
#define DISPID_STATBAR_SHOWTOOLTIPS												22
#define DISPID_STATBAR_SIMPLEMODE													23
#define DISPID_STATBAR_SIMPLEPANEL												DISPID_VALUE
#define DISPID_STATBAR_SUPPORTOLEDRAGIMAGES								24
#define DISPID_STATBAR_USESYSTEMFONT											25
#define DISPID_STATBAR_VERSION														26
// fingerprint
#define DISPID_STATBAR_APPID															500
#define DISPID_STATBAR_APPNAME														501
#define DISPID_STATBAR_APPSHORTNAME												502
#define DISPID_STATBAR_BUILD															503
#define DISPID_STATBAR_CHARSET														504
#define DISPID_STATBAR_ISRELEASE													505
#define DISPID_STATBAR_PROGRAMMER													506
#define DISPID_STATBAR_TESTER															507
// methods
#define DISPID_STATBAR_ABOUT															DISPID_ABOUTBOX
#define DISPID_STATBAR_GETBORDERS													27
#define DISPID_STATBAR_HITTEST														28
#define DISPID_STATBAR_LOADSETTINGSFROMFILE								29
#define DISPID_STATBAR_REFRESH														DISPID_REFRESH
#define DISPID_STATBAR_SAVESETTINGSTOFILE									30
#define DISPID_STATBAR_FINISHOLEDRAGDROP									31


// _IStatusBarEvents
// methods
#define DISPID_STATBARE_CLICK															DISPID_VALUE
#define DISPID_STATBARE_CONTEXTMENU												1
#define DISPID_STATBARE_DBLCLICK													2
#define DISPID_STATBARE_DESTROYEDCONTROLWINDOW						3
#define DISPID_STATBARE_MCLICK														4
#define DISPID_STATBARE_MDBLCLICK													5
#define DISPID_STATBARE_MOUSEDOWN													6
#define DISPID_STATBARE_MOUSEENTER												7
#define DISPID_STATBARE_MOUSEHOVER												8
#define DISPID_STATBARE_MOUSELEAVE												9
#define DISPID_STATBARE_MOUSEMOVE													10
#define DISPID_STATBARE_MOUSEUP														11
#define DISPID_STATBARE_OLEDRAGDROP												12
#define DISPID_STATBARE_OLEDRAGENTER											13
#define DISPID_STATBARE_OLEDRAGLEAVE											14
#define DISPID_STATBARE_OLEDRAGMOUSEMOVE									15
#define DISPID_STATBARE_OWNERDRAWPANEL										16
#define DISPID_STATBARE_PANELMOUSEENTER										17
#define DISPID_STATBARE_PANELMOUSELEAVE										18
#define DISPID_STATBARE_RCLICK														19
#define DISPID_STATBARE_RDBLCLICK													20
#define DISPID_STATBARE_RECREATEDCONTROLWINDOW						21
#define DISPID_STATBARE_RESIZEDCONTROLWINDOW							22
#define DISPID_STATBARE_SETUPTOOLTIPWINDOW								23
#define DISPID_STATBARE_TOGGLEDSIMPLEMODE									24
#define DISPID_STATBARE_XCLICK														25
#define DISPID_STATBARE_XDBLCLICK													26


// IStatusBarPanel
// properties
#define DISPID_SBP_ALIGNMENT															1
#define DISPID_SBP_BORDERSTYLE														2
#define DISPID_SBP_CONTENT																3
#define DISPID_SBP_CURRENTWIDTH														4
#define DISPID_SBP_ENABLED																5
#define DISPID_SBP_FORECOLOR															6
#define DISPID_SBP_HICON																	7
#define DISPID_SBP_INDEX																	8
#define DISPID_SBP_MINIMUMWIDTH														9
#define DISPID_SBP_PANELDATA															10
#define DISPID_SBP_PARSETABS															11
#define DISPID_SBP_PREFERREDWIDTH													12
#define DISPID_SBP_RIGHTTOLEFTTEXT												13
#define DISPID_SBP_TEXT																		DISPID_VALUE
#define DISPID_SBP_TOOLTIPTEXT														14
// methods
#define DISPID_SBP_GETRECTANGLE														15


// IStatusBarPanels
// properties
#define DISPID_SBPS_ITEM																	DISPID_VALUE
#define DISPID_SBPS__NEWENUM															DISPID_NEWENUM
// methods
#define DISPID_SBPS_ADD																		1
#define DISPID_SBPS_COUNT																	2
#define DISPID_SBPS_REMOVE																3
#define DISPID_SBPS_REMOVEALL															4


// IOLEDataObject
// properties
// methods
#define DISPID_ODO_CLEAR																	1
#define DISPID_ODO_GETCANONICALFORMAT											2
#define DISPID_ODO_GETDATA																3
#define DISPID_ODO_GETFORMAT															4
#define DISPID_ODO_SETDATA																5
#define DISPID_ODO_GETDROPDESCRIPTION											6
#define DISPID_ODO_SETDROPDESCRIPTION											7
